SystemAudioPlayer/impl/AudioPlayer.cpp

```diff
@@ -87,30 +90,50 @@ AudioPlayer::~AudioPlayer()
 
 void AudioPlayer::Init(SAPEventCallback *callback)
 {
+    std::unique_lock<std::mutex> lock(m_eventMutex);
     SAPLOG_INFO("SAP: AudioPlayer Init\n");
     if(!gst_is_initialized())
         gst_init(NULL,NULL);
+
     m_main_loop_thread = g_thread_new("BusWatch", (void* (*)(void*)) event_loop, NULL);
+    while(!m_eventCondition.wait_for(lock, std::chrono::seconds(1), [] {
+        return AudioPlayer::m_isLoopStarted;
+    }));
+
     m_callback = callback;
     Soc_Initialize();
 }
 
-void AudioPlayer::DeInit()
-{   
-    SAPLOG_INFO("SAP: AudioPlayer DeInit\n");
-    if(g_main_loop_is_running(m_main_loop))
-        g_main_loop_quit(m_main_loop);
-    g_thread_join(m_main_loop_thread);
-    Soc_Deinitialize();
-    SAPLOG_INFO("SAP: AudioPlayer DeInit last\n");
-}
-
 void AudioPlayer::event_loop()
 {
+    if (!m_main_loop && m_isLoopStarted)
+        return;
+
     m_main_loop = g_main_loop_new(NULL, false);
+    g_idle_add([] (gpointer data) -> gboolean {
+        std::unique_lock<std::mutex> lock(AudioPlayer::m_eventMutex);
+        AudioPlayer::m_isLoopStarted = true;
+        AudioPlayer::m_eventCondition.notify_one();
+        return G_SOURCE_REMOVE;
+    }, nullptr);
     g_main_loop_run(m_main_loop);
 }
 
+void AudioPlayer::DeInit()
+{
+    SAPLOG_INFO("SAP: AudioPlayer DeInit\n");
```

Comment:
In the below scenario, we might end up in deadlock
T1 : waiting for T2 to do g_main_loop_run()
T2 : is delayed (or hasn't started g_main_loop_run() yet)
T3 : is calling DeInit()

To unblock this case, having `m_isLoopStarted = true` in DeInit() would help (would force T1 to come out of wait, would force T2 to prematurely return without starting g_main_loop_run())

Reply:
added required changes